BCnEnc.Net/Decoder/RawDecoder.cs

```diff
@@ -1,139 +1,238 @@
-        public Rgba32[] Decode(ReadOnlySpan<byte> data, int pixelWidth, int pixelHeight)
-        {
-            var output = new Rgba32[pixelWidth * pixelHeight];
-            for (var i = 0; i < output.Length; i++)
-            {
-                output[i].R = data[i * 3];
-                output[i].G = data[i * 3 + 1];
-                output[i].B = data[i * 3 + 2];
-                output[i].A = 255;
-            }
-
-            return output;
-        }
-    }
-
-    /// <summary>
-    /// A class to decode data to RGBA components.
-    /// </summary>
-    public class RawRgbaDecoder : IRawDecoder
-    {
-        /// <summary>
-        /// Decode the data to color components.
-        /// </summary>
-        /// <param name="data">The data to decode.</param>
-        /// <param name="pixelWidth">The width of the image in pixels.</param>
-        /// <param name="pixelHeight">The height of the image in pixels.</param>
-        /// <returns>The decoded color components.</returns>
-        public Rgba32[] Decode(ReadOnlySpan<byte> data, int pixelWidth, int pixelHeight)
-        {
-            var output = new Rgba32[pixelWidth * pixelHeight];
-            for (var i = 0; i < output.Length; i++)
-            {
-                output[i].R = data[i * 4];
-                output[i].G = data[i * 4 + 1];
-                output[i].B = data[i * 4 + 2];
-                output[i].A = data[i * 4 + 3];
-            }
-
-            return output;
-        }
-    }
+	internal interface IRawDecoder
+	{
+		Rgba32[] Decode(ReadOnlyMemory<byte> data, int pixelWidth, int pixelHeight, OperationContext context);
+	}
+
+	/// <summary>
+	/// A class to decode data to R components.
+	/// </summary>
+	public class RawRDecoder : IRawDecoder
+	{
+		private readonly bool redAsLuminance;
+
+		/// <summary>
+		/// Create a new instance of <see cref="RawRDecoder"/>.
+		/// </summary>
+		/// <param name="redAsLuminance">If the decoded component should be used as the red component or luminence.</param>
+		public RawRDecoder(bool redAsLuminance)
+		{
+			this.redAsLuminance = redAsLuminance;
+		}
+
+		/// <summary>
+		/// Decode the data to color components.
+		/// </summary>
+		/// <param name="data">The data to decode.</param>
+		/// <param name="pixelWidth">The width of the image in pixels.</param>
+		/// <param name="pixelHeight">The height of the image in pixels.</param>
+		/// <param name="context">The context of the current operation.</param>
+		/// <returns>The decoded color components.</returns>
+		public Rgba32[] Decode(ReadOnlyMemory<byte> data, int pixelWidth, int pixelHeight, OperationContext context)
+		{
+			var output = new Rgba32[pixelWidth * pixelHeight];
+
+			if (context.IsParallel)
+			{
+				var options = new ParallelOptions
+				{
+					MaxDegreeOfParallelism = context.TaskCount,
+					CancellationToken = context.CancellationToken
+				};
+				Parallel.For(0, pixelWidth * pixelHeight, options, i =>
```

Comment:
Doing this pixel by pixel is probably a few times slower than non-parallel. In the raw decoders the parallelization should be in chunks. Maybe line by line.

Reply:
Or might not need  parallelzation in the raw decoders at all. The speedup is probably very marginal.

Reply:
So removing parallelity from the raw decoders and mentioning that in the IsParallel property when it doesn't have an effect. Got it.